### PR TITLE
f16,f32,.. don't convert int arrays + handle complex

### LINKED ISSUE #2225: should f16,f32, ... convert integer or boolean arrays?
I think it would make sense for `f16` etc to convert only arrays of float and leave integers be. 
This is an example of current behavior that should be changed:
```julia
julia> y = Flux.onehotbatch([1,2,3], 1:3)
3×3 OneHotMatrix(::Vector{UInt32}) with eltype Bool:
 1  ⋅  ⋅
 ⋅  1  ⋅
 ⋅  ⋅  1

julia> y |> f16
3×3 Matrix{Float16}:
 1.0  0.0  0.0
 0.0  1.0  0.0
 0.0  0.0  1.0
```
This is instead an error that could be avoided by preserving integer arrays
```julia
julia> batch = (rand(2, 2), Flux.onehotbatch([1,2,3], 1:3));

julia> batch |> f16
ERROR: TypeError: in OneHotArray, in T, expected T<:Integer, got Type{Float16}
Stacktrace:
  [1] OneHotArrays.OneHotArray(indices::Vector{Float16}, L::Int64)
    @ OneHotArrays ~/.julia/packages/OneHotArrays/T3yiq/src/array.jl:17
  [2] adapt_structure
    @ ~/.julia/packages/OneHotArrays/T3yiq/src/array.jl:143 [inlined]
  [3] adapt
    @ ~/.julia/packages/Adapt/xviDc/src/Adapt.jl:40 [inlined]
  [4] Fix1
    @ ./operators.jl:1108 [inlined]
  [5] ExcludeWalk
    @ ~/.julia/packages/Functors/8Lxin/src/walks.jl:92 [inlined]
  [6] (::Functors.CachedWalk{Functors.ExcludeWalk{Functors.DefaultWalk, Base.Fix1{typeof(Adapt.adapt), Flux.FluxEltypeAdaptor{Float16}}, typeof(Functors.isleaf)}, Functors.NoKeyword})(::Function, ::OneHotArrays.OneHotMatrix{UInt32, Vector{UInt32}})
    @ Functors ~/.julia/packages/Functors/8Lxin/src/walks.jl:132
  [7] recurse
    @ ~/.julia/packages/Functors/8Lxin/src/maps.jl:6 [inlined]
  [8] map
    @ ./tuple.jl:274 [inlined]
  [9] _map
    @ ~/.julia/packages/Functors/8Lxin/src/walks.jl:1 [inlined]
 [10] DefaultWalk
    @ ~/.julia/packages/Functors/8Lxin/src/walks.jl:62 [inlined]
 [11] ExcludeWalk
    @ ~/.julia/packages/Functors/8Lxin/src/walks.jl:92 [inlined]
 [12] CachedWalk
    @ ~/.julia/packages/Functors/8Lxin/src/walks.jl:132 [inlined]
 [13] fmap
    @ ~/.julia/packages/Functors/8Lxin/src/maps.jl:7 [inlined]
 [14] #fmap#26
    @ ~/.julia/packages/Functors/8Lxin/src/maps.jl:18 [inlined]
 [15] fmap
    @ ~/.julia/packages/Functors/8Lxin/src/maps.jl:10 [inlined]
 [16] _paramtype
    @ ~/.julia/packages/Flux/Nzh8J/src/functor.jl:274 [inlined]
 [17] f16
    @ ~/.julia/packages/Flux/Nzh8J/src/functor.jl:318 [inlined]
 [18] |>(x::Tuple{Matrix{Float64}, OneHotArrays.OneHotMatrix{UInt32, Vector{UInt32}}}, f::typeof(f16))
    @ Base ./operators.jl:907
 [19] top-level scope
```
So I guess my suggestion is to change
https://github.com/FluxML/Flux.jl/blob/90f207d6e8b4a8939777809f8604ea15d44ab833/src/functor.jl#L272
to 
```julia
Adapt.adapt_storage(::FluxEltypeAdaptor{T}, x::AbstractArray{<: AbstractFloat}) where {T <: AbstractFloat} = convert(AbstractArray{T}, x)
```
although I'm not sure this is all it is needed.

### PR DESCRIPTION
**Pull request by CarloLucibello**
Close #2225 + support `Complex` types

### PR Checklist

- [x] Tests are added
- [x] Entry in NEWS.md
- [x] Documentation, if applicable

### CODE REVIEW

File: src/functor.jl [hunk 1]
```diff
@@ -269,16 +269,24 @@ ChainRulesCore.@non_differentiable check_use_cuda()
...
+  convert(AbstractArray{T}, x)
+_paramtype(::Type{T}, x::AbstractArray{<:Complex{<:AbstractFloat}}) where {T<:AbstractFloat} = 
+  convert(AbstractArray{Complex{T}}, x)
 
 """
     f32(m)
 
 Converts the `eltype` of model's parameters to `Float32` (which is Flux's default).
-Recurses into structs marked with [`@functor`](@ref).
+Recurses into structs marked with [`@functor`](@ref). Preserves integer arrays.
```
Comment (darsnack) [inline src/functor.jl:289]: Suggestion:
```
Converts the `eltype` of model's *floating point* parameters to `Float32` (which is Flux's default).
Recurses into structs marked with [`@functor`](@ref).

File: NEWS.md [hunk 2]
```diff
@@ -1,5 +1,9 @@
 # Flux Release Notes
 
+## v0.13.16
+* `f16, f32, f64` now preserve integer arrays.
+* `f16, f32, f64` can now handle `Complex` arrays.
```
Comment (darsnack) [inline NEWS.md:5]:
```suggestion
* `f16, f32, f64` now specifically target floating point arrays (i.e. integers arrays and other types are preserved).
* `f16, f32, f64` can now handle `Complex{<:AbstractFloat}` arrays.
```

**Review by darsnack — COMMENTED** [review]
Just to be more clear about what the code does.

File: src/functor.jl [hunk 3]
```diff
@@ -269,16 +269,24 @@ ChainRulesCore.@non_differentiable check_use_cuda()
...
+# fastpath for arrays
+_paramtype(::Type{T}, x::AbstractArray{<:AbstractFloat}) where {T<:AbstractFloat} = 
+  convert(AbstractArray{T}, x)
+_paramtype(::Type{T}, x::AbstractArray{<:Complex{<:AbstractFloat}}) where {T<:AbstractFloat} = 
+  convert(AbstractArray{Complex{T}}, x)
 
 """
     f32(m)
 
 Converts the `eltype` of model's parameters to `Float32` (which is Flux's default).
```
Comment (CarloLucibello) [inline src/functor.jl:288]:
```suggestion
Converts the `eltype` of model's *floating point* parameters to `Float32` (which is Flux's default).
```

File: src/functor.jl (hunk 1 above)
Comment (CarloLucibello) [inline src/functor.jl:289]:
```suggestion
Recurses into structs marked with [`@functor`](@ref).
```

**Comment by CarloLucibello** [comment]
should be ready

**Review by darsnack — APPROVED** [review]

**Pull request merged**